Graphical UI for Network Inventory

benthomasson opened the pull request:
##### SUMMARY

Adds a graphical interface for interacting with hosts and groups with customizations specifically for networking workflows. 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

 - API: awx/network_ui
 - UI: awx/ui/client/src/network_ui

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 1.0.4.456
```


##### ADDITIONAL INFORMATION

See awx/network_ui/CONTRIBUTING.md and awx/network_ui/designs/README.md for more details on this feature.

benthomasson committed:
Adds configuration for the network-ui websocket

* Configures NGINX for the network-ui websocket.
* Configures supervisor.conf for network_ui websocket.

benthomasson committed:
Adds callback plugin for network_ui

The callback plugin for the network UI adds real-time
event streaming to the canvas from Ansible events.

benthomasson committed:
Begins network-ui prototype integration into Tower UI.

* Moves network ui into a directive
* Adds awxNet prefix to network ui directives
* Adds a module to integrate the stand alone network UI with
    Tower UI.
* Adds reconnectingwebsocket to webpack bundle
* Adds configuration for webpack
* Moves ngTouch and hamsterjs to webpack vendor bundle
* Moves angular to network UI vendor bundle
* Adds ui-router dependency
* Changes CSS to BEM style
* Adds unique id sequences for devices and links on Topology and interfaces on Device
* Adds group widget with move, resize, delete, and edit label support

benthomasson committed:
Adds JSON and YAML export of the network topology.

Adds views that export the entire network topology as JSON and YAML.

benthomasson committed:
Imports prototype from ansible-network-ui

The ansible-network-ui prototype project builds a standalone Network UI
outside of Tower as its own Django application. The original prototype
code is located here:
https://github.com/benthomasson/ansible-network-ui.

The prototype provides a virtual canvas that supports placing
networking devices onto 2D plane and connecting those devices together
with connections called links.  The point where the link connects
to the network device is called an interface.  The devices, interfaces,
and links may all have their respective names.  This models physical
networking devices is a simple fashion.

The prototype implements a pannable and zoomable 2D canvas in using SVG
elements and AngularJS directives.   This is done by adding event
listeners for mouse and keyboard events to an SVG element that fills the
entire browser window.

Mouse and keyboard events are handled in a processing pipeline where
the processing units are implemented as finite state machines that
provide deterministic behavior to the UI.

The finite state machines are built in a visual way that makes
the states and transitions clearly evident.  The visual tool for
building FSM is located here:
https://github.com/benthomasson/fsm-designer-svg.   This tool
is a fork of this project where the canvas is the same.  The elements
on the page are FSM states and the directional connections are called
transitions.   The bootstrapping of the FSM designer tool and
network-ui happen in parallel.  It was useful to try experiemental
code in FSM designer and then import it into network-ui.

The FSM designer tool provides a YAML description of the design
which can be used to generate skeleton code and check the implementation
against the design for discrepancies.

Events supported:

* Mouse click
* Mouse scroll-wheel
* Keyboard events
* Touch events

Interactions supported:

* Pan canvas by clicking-and-dragging on the background
* Zooming canvas by scrolling mousewheel
* Adding devices and links by using hotkeys
* Selecting devices, interaces, and links by clicking on their icon
* Editing labels on devices, interfaces, and links by double-clicking on
  their icon
* Moving devices around the canvas by clicking-and-dragging on their
  icon

Device types supported:

* router
* switch
* host
* racks

The database schema for the prototype is also developed with a visual
tool that makes the relationships in the snowflake schema for the models
quickly evident.  This tool makes it very easy to build queries across
multiple tables using Django's query builder.

See: https://github.com/benthomasson/db-designer-svg

The client and the server communicate asynchronously over a websocket.
This allows the UI to be very responsive to user interaction since
the full request/response cycle is not needed for every user
interaction.

The server provides persistence of the UI state in the database
using event handlers for events generated in the UI.  The UI
processes mouse and keyboard events, updates the UI, and
generates new types of events that are then sent to the server
to be persisted in the database.

UI elements are tracked by unique ids generated on the client
when an element is first created.  This allows the elements to
be correctly tracked before they are stored in the database.

The history of the UI is stored in the TopologyHistory model
which is useful for tracking which client made which change
and is useful for implementing undo/redo.

Each message is given a unique id per client and has
a known message type.  Message types are pre-populated
in the MessageType model using a database migration.

A History message containing all the change messages for a topology is
sent when the websocket is connected.  This allows for undo/redo work
across sessions.

This prototype provides a server-side test runner for driving
tests in the user interface.  Events are emitted on the server
to drive the UI.  Test code coverage is measured using the
istanbul library which produces instrumented client code.
Code coverage for the server is is measured by the coverage library.

The test code coverage for the Python code is 100%.

benthomasson committed:
Check for missing environ in request

benthomasson committed:
Adds support for multiple view modes

Adds mulitple view modes based on zoom-level.  This allows for easy
drilling into a device for more detail or zooming-out for a overview.

* Adds support for multi-site and device modes
* Adds icons to remote device in device detail
* Adds site widget
* Adds link between sites
* Adds toolboxes for inventory, site, and applications
* Adds rack mode
* Adds UI for adding processes to devices
* Adds copy and paste support
* Adds streams

benthomasson committed:
Adds a type field to group to support racks and sites

* Add type support to Group

benthomasson committed:
Adds streams and processes for application design

Adds application level streams and process widgets to
model applications that run on networking devices or hosts.

* Changes Application to Process
* Adds StreamCreate and ProcessCreate messages
* Adds process id sequence to device
* Add serializers for streams and processes

benthomasson committed:
Adds toolbox to network UI

* Calls API to get inventory
* Adds CopySite message
* Adds Toolbox and ToolboxItem model design
* Add Toolbox and ToolboxItem tables
* Sends toolbox items to client from server on connect

benthomasson committed:
Adds facts processing for ansible_net_neighbors

* Adds logic for consuming ansible_net_neighbors facts

This consumes facts emitted from Ansible over a websocket to
Tower.  This allows consumers in network to process the facts and
emit messges to the network UI.  This requires a special callback
plugin to run in Tower to emit the messages into the websocket using
the python websocket-client library.

benthomasson committed:
Adds task status on the device

* Show task status on device for now

This shows the status of the last few tasks run on a device as a
green/red circle on the device icon.  This data live updates
from data emitted over the websocket.

benthomasson committed:
Improves FSM design and adds tools to diff design and implementation.

* Resolves conflicts between designs and implementation
* Adding fsm_diff to network_ui/tools
* Add extract.js for FSM toolchain to network_ui

jaredevantabor committed:
Adds network UI shell wrapper

* Adds networking icons, state, and shell
* Adds network UI to the network UI shell.
* Removes jquery as a dependency of network-ui
* Fills the entire viewport with the network canvas and
    makes header panel and the right panel overlay on
    top of it

benthomasson committed:
Adds Red Hat copyright notice

benthomasson committed:
Adds devserver support

* Adds support for webpack devserver
* Enable istanbul on network UI
* Enable capture and replay tests on the network ui
* Normalize mouse wheel events
* Fix missing trailing slash on hosts API
* Add Export YAML button

jaredevantabor committed:
Adds host detail panel UI and improves toolbox UI

* Changes Layers' panel's default setting to not expanded
* Adds OffScreen2 state to handle the case where a toolbox is both offscreen and disabled
* Adds a collapsed view of the toolbox, as well as a model for ActionIcons
    which is a model whose purpose is to connect the button FSM with the
    chevron icons that are used on the toolbox.
* Adds action-icon directive
* Enables/disables the icons if they're not shown
* Fixes initial state of the toolboxes
* Creates context menu and context menu buttons in the network UI
* Adds extra vars to details panel on left hand side

benthomasson committed:
Adds content to CONTRIBUTING.md

* Adds SVG intro to CONTRIBUTING.md
* Add FSM intro
* Add rendered images of the FSM designs
* Adding example
* Adding links
* Adds details about the FSM design workflows
* Adds FSM state docs
* Adds event handler docs
* Adds details about FSMController
* Adds example of making an FSMController
* Adds details about messages, models, and message passing
* Adds models and messages to CONTRIBUTING.md
* Adds example to widget development
* Adds detail to the widget development example
* Add message type definitions

benthomasson committed:
Adds explicit channels between FSMs to add in tracing message flows.

* Adds channels between FSMs
* Adds FSMTrace model
* Adds FSMTrace storage and download

Channels between FSMs make the processing pipeline delegation explicit
and allow for better instrumentation to trace the state of the entire
pipeline including FSM state transitions and message flow through
the pipeline.  This feature is not turned on by default and is
only necessary for debugging or certain kinds of testing.

benthomasson committed:
Reorganizes the network-ui code under awx/client/src/network-ui

* Moves network UI source to awx/client/src/network-ui
* Moves network ui partials to awx/ui/client/network-ui
* Renames widgets with suffix partial.svg
* Updates directives to use bundled partials
* Uses ~network-ui for loading UI component

jaredevantabor committed:
Improves host details panel UI

* Hooks up the first two context menu buttons
* Makes the rename and details menu show up
    wherever the user's cursor's location
* Adds TopologyInventory and DeviceHost tables
* Adds design for host_id on the Device table
* Adds migrations for TopologyInventory
* Adds host_id to Device table
* Adds inventory_id and host_id tracking
* Auto-closes the right hand panel if focus is directed to the canvas.
* Retrieves the host details on inventory load.
* Adds back support for inventory and host_id tracking
* Adds host icon
* Changes rack icon to new icon
* Site icon replacement
* Fixes host icon "hitbox", and adding debug and construction
* Adds construction and debug lines for switch, router, rack, and site
* Adds some error handling for REST calls, as well as alert on
    host detail panel.

benthomasson committed:
Removes early experiments from network UI.

* Removing unused widgets app
* Removes dead code for rack from move
* Removes experimental touch support
* Removes unused Group corners function
* Removes experimental tables app
* Removes stencil and layers
* Removes status light and task status
* Removes configuration
* Removes unused PasteGroup message and handler
* Removes unused inventory clip path partial
* Removes old recordings
* Removes unused table messages
* Removes unused Task and Configuration models

benthomasson committed:
Adds network UI test framework

This adds a test framework to drive UI tests from the client
instead of injecting events from the websocket.  Tests consist
of a pair of snapshots (before and after the test) and
a list of UI events to process.  Tests are run using a FSM
in the client that controls the resetting of state to the snapshot,
injecting the events into the UI, recording test coverage,
and reporting tests to the server.

* Adds design for event trace table
* Adds design for a coverage tracking table
* Adds models for EventTrace and Coverage
* Adds trace_id to recording messages
* Adds design for TopologySnapshot table
* Adds order to TopologySnapshot table
* Adds TopologySnapshot table
* Adds Snapshot message when recordings are started and stoppped
* Adds models for tracking test cases and test results
* Adds designs for a test runner FSM
* Updates test management commands with new schema
* Adds download recording button
* Adds models to track tests
* Adds ui test runner
* Adds id and client to TestResult design
* Adds id and client to TestResult
* Update message types
* Stores test results and code coverage from the test runner
* Adds tool to generate a test coverage report
* Adds APIs for tests and code coverage
* Adds per-test-case coverage reports
* Breaks out coverage for loading the modules from the tests
* Re-raises server-side errors
* Captures errors during tests
* Adds defaults for host name and host type
* Disables test FSM trace storage
* Adds support for sending server error message to the client
* Resets the UI flags, history, and toolbox contents between tests
* Adds istanbul instrumentation to network-ui

benthomasson committed:
Fixes #945 by removing the network_ui plugin

benthomasson committed:
Updates pipeline and FSM design and development tools

* Updates pipeline and FSM design for 3.4 features:
    group and read/write design features.
* Adds tool to copy layout from existing design
* Adds pipeline design

benthomasson committed:
Adds search field and jump-to a device UI.

Adds a search field in the network UI and a jump-to level menu. This
allows users to quickly find a device on the canvas or jump to a
certain mode/zoom-level.

Adds animation to smooth out the transition from the current viewport
to a viewport centered on the searched for device or zoom-level.

* Adds animation FSM and changes the 0 hot key to use it
* Adds jump to animation
* Adds search bar type ahead
* Adds jump animation to search and jump-to menus
* Adds keybinding FSM
* Updates the dropdown when devices are added/edit/removed
* Highlights the searched for host

jaredevantabor committed:
Adds editing to the host details form.

* Adds editing host vars, description, and name to the host details
form.
* Adds details panel FSM and updates to keybindings FSM

benthomasson committed:
Adds breadcrumb to network UI

* Adds functionality for breadcrumb
* Changes the site icon to stay the same size

benthomasson committed:
Cleans up feature set for 3.3

This removes the experimental features that were not selected
for 3.3 release.

* Removes dpath requirement
* Removes generated action_plugins
* Removes network UI v1 api
* Removes unused network management commands
* Removes network UI CLI client
* Removes templates
* Removes unused DataBinding models
* Removes obsolete test
* Removes unused admin and tests
* Removes experimental UndoPersistence, RedoPersistence, and auto-layout functions
* Removes API endpoints for network visualization
* Removes unused consumer routes
* Removes group, site, and rack features for 3.3
* Removes unused tables controller
* Removes undo/redo
* Removes group code and scale checks

jaredevantabor committed:
Cleans up network UI code for 3.3

This removes features that were not selected for 3.3.

* Removes breadcrumb
* Removes "Jump To" panel and some of the hotkey panel items
* Removes Buttons in favor of Action Dropdown
* Removes chevrons
* Removes ActionIcon model
* Removes the Rename button on the context menu
* Makes details panel readonly
* Adds expand modal for extra vars
* Adds inventory copy function back to inventory list
* Sets cursor to visible
* Adds hide_menus
* Adds fix for mice that return large mousewheel deltas

benthomasson committed:
Adds inventory tracking and templating to network UI groups and hosts.

* Adds group_id to Group table
* Adds inventory_group_id to Group table
* Adds creation of inventory hosts and groups from the network UI
* Changes network UI variables to be under awx key
* Fixes variables initial value
* Adds group membership association/disassociation
* Removes items from the inventory toolbar when loaded by a snaphot
* Adds nunjucks dependency to package.json
* Adds templating to hosts
* Adds templating for racks
* Adds site templating
* Adds group associations for sites
* Squashes migrations for network_ui
* Flake8 migrations
* Changes reserved field type to device_type, group_type, and process_type
* Allows blank values for all CharFields in network_ui models
* Changes reserved field type to device_type, group_type, and process_type

benthomasson committed:
Fixes double import of instanceGroups app

matburt commented:
I'm going to start looking at this this afternoon but I definitely want to see the commit numbers to be consolidated from ~400 to maybe a dozen or so.

Hopefully we can pare this down from ~30k lines also

benthomasson commented:
I moved the client code to other repos.   The python code in this PR is not too big.

```$ find . -name "*.py" | xargs -exec wc -l
   103 ./serializers.py
  1006 ./consumers.py
   616 ./v2_api_views.py
     6 ./utils.py
    28 ./urls.py
    20 ./routing.py
   212 ./models.py
    74 ./v2_api_urls.py
   101 ./v2_api_access.py
     2 ./__init__.py
    61 ./tools/coverage_report.py
   138 ./views.py
   557 ./migrations/0001_squashed_0036_auto_20180223_0119.py
   130 ./migrations/0003_auto_20180302_1401.py
     0 ./migrations/__init__.py
    40 ./migrations/0004_auto_20180304_2122.py
    36 ./migrations/0002_auto_20180301_2039.py
    33 ./messages.py
   129 ./v2_api_serializers.py
  3292 total```

benthomasson commented:
The PR is mostly JS and SVG content.

From `awx/awx/ui/client/src/network-ui`:

``` 
 find . -name "*.js" | xargs -exec wc -l
...
   133 ./hotkeys.fsm.js
   153 ./tables.ui.controller.js
 10231 total
```
```
find . -name "*.svg" | xargs -exec wc -l
...
     83 ./router.partial.svg
     12 ./chevron_right.partial.svg
   4612 total
```

benthomasson committed:
Adds creating links from host vars from inventory

* Adds creating links from host vars
* Returns devices to the toolbox after they are removed from the canvas

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+        client_id = message.get('client')
+        assert client_id is not None, "No client_id"
+        data = json.loads(message['text'])
+        if isinstance(data[1], list):
```
Subscript not in an exception block or without explicitly checking length. Also more of that below.

benthomasson replied:
I'll add a parse_message_text function that does this properly.

benthomasson replied:
I'll rewrite this to properly parse the messages.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+                logger.warning("Unsupported message %s", message['msg_type'])
+
+
+persistence = _Persistence()
```
I'd rather see this instance being created where it is actually used in `routing.py`

benthomasson replied:
Like this?
```route("persistence", Persistence().handle),```

wwitzel3 replied:
Sure, or just create persistence inside routing

wwitzel3 commented on awx/network_ui_test/consumers.py:
```diff
@@ -0,0 +1,205 @@
+import json
+# Connected to websocket.connect
+
+HISTORY_MESSAGE_IGNORE_TYPES = ['DeviceSelected',
```
I feel like some of these would be easier to scan/read as fewer lines.

wwitzel3 commented on awx/network_ui_test/migrations/0002_auto_20180306_2243.py:
```diff
@@ -0,0 +1,34 @@
+from django.db import migrations
+
+
+results = ['passed',
```
Easier to read as fewer lines.

wwitzel3 commented on awx/network_ui_test/tools/coverage_report.py:
```diff
@@ -0,0 +1,61 @@
+            os.mkdir(test['name'])
+        with open(test['name'] + "/coverage.json", 'w') as f:
+            f.write(requests.get(server + test['coverage'], verify=False).text)
+
```
So


Much


Space

wwitzel3 replied:
Don't worry about this, just noticed it was a testing tool.

benthomasson replied:
I thought that was a flake8 violation.  I'm not sure why that much space was allowed.

wwitzel3 commented on awx/network_ui_test/consumers.py:
```diff
@@ -0,0 +1,205 @@
+            except NetworkUIException, e:
+                Group("client-%s" % client_id).send({"text": json.dumps(["Error", str(e)])})
+                raise
+            except Exception, e:
```
Same as above. Not needed.

benthomasson replied:
Thanks, I'll remove it.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+import json
+# Connected to websocket.connect
+
+HISTORY_MESSAGE_IGNORE_TYPES = ['DeviceSelected',
```
I would find this easier to read as less lines.

benthomasson replied:
I'll remove this as it is no longer needed.  The undo/redo feature was removed for this release.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+    data = urlparse.parse_qs(message.content['query_string'])
+    inventory_id = parse_inventory_id(data)
+    topology_ids = list(TopologyInventory.objects.filter(inventory_id=inventory_id).values_list('topology_id', flat=True))
+    topology_id = 0
```
This code assumes that a valid `topology_id` will never be 0, it would be slightly better to set it to `None` and check `if topology_id is not None` on 203

benthomasson replied:
Yes,  the database does not restrict ids to positive numbers.  I'll change this.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+              .filter(device__topology_id=topology_id)
+              .values()):
+        interfaces[i['device_id']].append(i)
+    devices = list(Device.objects
```
This would be easier to scan/read as 1 line

benthomasson replied:
I agree.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,277 @@
+            except Exception, e:
+                Group("client-%s" % client_id).send({"text": json.dumps(["Error", "Server Error"])})
+                raise
+            except BaseException, e:
```
As of 2.5 `Exception` inherits from `BaseException` so this `except` can be removed. Also I don't see anywhere you raise a `NetworkUIException` and there is no difference in how the exception is handled. I would delete it. And up date this block to just handling `Exception` OR better, to handle the known exceptions that can come out of calling `handle`

benthomasson replied:
I will reduce this to just the except Exception clause.

benthomasson committed:
Resetting migrations for network_ui

benthomasson committed:
Makes changes suggested by wwitzel3's review

benthomasson committed:
Adds CONTRIBUTING docs

mabashian commented on awx/ui/client/src/inventories-hosts/inventories/inventory.list.js:
```diff
@@ -93,6 +93,13 @@ export default ['i18n', function(i18n) {
+                ngClick: 'goToGraph(inventory)',
+                awToolTip: i18n._('Network Visualization'),
+                dataPlacement: 'top',
+                ngShow: '!inventory.pending_deletion && inventory.summary_fields.user_capabilities.edit'
```
OK so just to confirm - there's no _read only_ view of the network topography?  A user must have edit permissions on the inventory to access this?

jaredevantabor replied:
Currently yes, but in the next few days I hope to change that so that the canvas has a readonly capability. More coming in the near future.

mabashian commented on awx/ui/client/src/inventories-hosts/inventories/list/inventory-list.controller.js:
```diff
@@ -84,6 +84,12 @@ function InventoriesList($scope,
+    
+    $scope.goToGraph = function(inventory){
+         $state.go('inventories.edit.networking', {inventory_id: inventory.id, inventory_name: inventory.name});
+        // let url = $state.href('inventories.edit.networking', {inventory_id: inventory.id, inventory_name: inventory.name});
```
Can probably remove if unused

mabashian commented on awx/ui/client/src/network-ui/network-details/details.controller.js:
```diff
@@ -0,0 +1,44 @@
+ 			HostsService.put(host).then(function(response){
+                $scope.saveConfirmed = true;
+                if(_.has(response, "data")){
+                    $scope.$parent.$broadcast('awxNet-hostUpdateSaved', response.data);
```
Talked with @jaredevantabor about mitigating the use of the $broadcast/$on pattern if possible.  I don't consider this to be something that would block the merge but is a nice thing to clean up _if possible_ in the future.

benthomasson replied:
Yes, I agree broadcasting messages should be avoided.   This should be changed to unicast a message to a specific listener.

jaredevantabor replied:
Yes, I'm planning on doing some cleanup in the near future to consolidate the scopes being used between the shell and the canvas itself. This will help mitigate the need for these $on/$broadcast/$emit events.

benthomasson committed:
Updates UI CONTRIBUTING for 3.3

* Fixes links in CONTRIBUTING.md
* Moves CONTRIBUTING and UI design files to ui/client/src/network-ui
* Adds README.md for network_ui/designs
* Updates design

benthomasson committed:
Adds a tabular view of the topology data

The traditional network engineer workflow includes a diagram, a
spreadsheet, and the CLI.  This adds an experimental view of the
network topology data in a spreadsheet like table view.

* Adds angular-xeditable dependency for tables view.
* Add data binding models
* Add message transformations from table to topology formats
* Adding dependencies for tables view

benthomasson committed:
Refactors network_ui_test out of network_ui

* Separates test messages from application messages
* Removes test runner and groups, processes, and streams from network_ui
* Adds network_ui_test
* Fixes routing for network_ui_test
* Removes coverage_report tool from network_ui
* Fixes network_ui_test test workflow
* Sets width and height of the page during tests

benthomasson committed:
Adds context menus for group, racks, and sites

* Adds context menu for a rack, and adding more error handling for
    items that don't exist in Tower
* Adds context menu for sites
* Adds handler for showing details for links and interfaces
* Fixes the removed "watchCollection" in order to update details panel
* Removes the context menu when changing the scale of the canvas
* Adds delete context menu button, as well as refactoring the delete
    functionality to the network.ui.controller.js
* Updates delete functionality to delete nested groups/devices
    if the current_scale is set to site or rack icons
* Adds context menu to a group
* Hides rack/site title in top left of group, as well as centering
    labels on all icons
* Moves the context menu off screen when disabling it
* Adds unique name to hosts, routers, switches, and groups
* Makes the names of host/switch/router/group SVG elements so they update when
    the user updates the name of the SVG element
* Removing svg buttons and adding new html toolbar
* Adds panel for Jump To feature, along with basic functionality
* Adds Key dropdown for hotkeys and adding browser refresh hotkey
* Adds breadcrumb bar and making adjustments after feedback with UX
* Rearrages panels and adding some resize logic
* Fixes z-index of key-panel  and jump-to panel
* Adds white background to text underneath icons
* Makes all icons blue
* Changes sizes and colors of icons. Also made icon text background white
* Adjusts sizes of rack and site icons within group boundary

benthomasson committed:
Adds an API for network UI, action plugins, and API client

* Adds a simple DRF API for network-ui
* Moves network_ui api to v1_api
* Uses BaseSerializer for networking v1 api
* Adds v2 of the network API
* Uses standard AWX base classes for the network UI API
* Adds canvas prefix to network UI api URL names
* Adds ansible action plugins for automating network UI workflows
* Adds python client for the networking visualization API

jaredevantabor committed:
Cleans up the Network UI after PR feedback and UX feedback

-removes stale commented-out lines
-makes "unknown" type devices smaller on canvas
-moves "unknown" type device title underneath icon
-removes collapsed inventory toolbox
-changes "Delete" to "Remove"
-removes the "Close" button for "Cancel" on details panel
-changing Remove color to red

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+# Copyright (c) 2017 Red Hat, Inc
+# In consumers.py
```
remove

benthomasson replied:
Removed.

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+from awx.network_ui.utils import transform_dict
+
+import json
+# Connected to websocket.connect
```
remove

benthomasson replied:
Removed

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+
+
+def parse_inventory_id(data):
+    inventory_id = data.get('inventory_id', ['null'])
```
How does this null string value differ from the `None` assignment below?

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+class Persistence(object):
+
+    def parse_message_text(self, message_text, client_id):
+        data = json.loads(message_text)
```
It might be helpful to show an example payload at the top of this class to make it easier to parse what this is doing.

benthomasson replied:
CONTRIBUTING.md has the ABNF for the message format.

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+    return inventory_id
+
+
+class Persistence(object):
```
I'm not sure what this class is for, might be worth adding a doc string to give an overview of purpose

wwitzel3 replied:
I think the name is what makes it confusing. This is really just a dynamic router for the different networking event, yeah?

matburt replied:
I think that makes sense... a clearer name and some comments in a docstring would go a long way here.

benthomasson replied:
I added a few short doc strings to help clear up the purpose.

wwitzel3 replied:
The name of this class is still far too generic. `NetworkUIEvents` maybe?

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+
+    def handle(self, message):
+        topology_id = message.get('topology')
+        assert topology_id is not None, "No topology_id"
```
I generally don't like `assert` statements in production code. I'd rather this raise an *Exception of something that could be handled below this method call or let it set some defaults such as what `message_type` below this does, which is `return`.   Same with `client_id`

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+        if handler is not None:
+            try:
+                handler(message_value, topology_id, client_id)
+            except Exception:
```
This is a very generalized exception, what could happen here explicitly?

wwitzel3 replied:
Still too general. Lets please handle the explicit exceptions.

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+        d.device_type = device['device_type']
+        d.host_id = device['host_id']
+        d.save()
+        (Topology.objects
```
Why is this wrapped up with parens?

benthomasson replied:
For style. So I can line up the .objects, .filter, and .update operations.

matburt commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,261 @@
+        Device.objects.filter(topology_id=topology_id, id=device['id']).update(name=device['name'])
+
+    def onInterfaceLabelEdit(self, interface, topology_id, client_id):
+        (Interface.objects
```
outer parens?

chrismeyersfsu commented on awx/network_ui_test/consumers.py:
```diff
@@ -0,0 +1,184 @@
+        inventory_id = int(inventory_id[0])
+    except ValueError:
+        inventory_id = None
+    if not inventory_id:
```
Non-obvious. `if inventory_id == 0` then you set `inventory_id` to `None`.

matburt replied:
I usually do the `if thing is None` or `if thing is not None` in these situations.

benthomasson replied:
How do you feel about this:  `inventory_id = int(inventory_id[0]) or None`?

matburt commented on awx/network_ui/designs/README.md:
```diff
@@ -0,0 +1,8 @@
+
```
Would actually like to see these moved under `docs/network_ui/` or something similar and the content here expanded upon.

matburt commented on awx/network_ui/static/network_ui/.gitignore:
```diff
@@ -0,0 +1,8 @@
+/bundle.js
```
move this to the gitignore at the top of the repository.

benthomasson replied:
Removed old file.

matburt commented on awx/network_ui/tools/Makefile:
```diff
@@ -0,0 +1,13 @@
+
+SERVER = "https://meganuke:8043"
```
This seems *very* specific... if these utilities are needed then lets generalize them at the top of the repository. If not then lets remove this Makefile.

Same for `requirements.txt` below.

benthomasson replied:
Removed old files.

chrismeyersfsu commented:
@benthomasson More meaningful migration names for migration other than the epoch.

chrismeyersfsu commented on awx/network_ui/serializers.py:
```diff
@@ -0,0 +1,69 @@
+                                                     'to_link__from_interface__name')
+
+
+def topology_data(topology_id):
```
I worry about not using DRF features (i.e. serializes). But particularly, pagination.

benthomasson replied:
This serializer is specifically for a single large data dump.   I've optimized it for small memory use per row.

matburt commented on awx/network_ui_test/consumers.py:
```diff
@@ -0,0 +1,184 @@
+# Copyright (c) 2018 Red Hat, Inc
+# In consumers.py
```
unneeded

benthomasson replied:
Removed

matburt commented on awx/network_ui_test/consumers.py:
```diff
@@ -0,0 +1,184 @@
+
+
+def parse_inventory_id(data):
+    inventory_id = data.get('inventory_id', ['null'])
```
This code looks duplicated from elsewhere.

matburt replied:
What is the purpose of this module `network_ui_test` ?

benthomasson replied:
Test framework.  Moving to external repo: https://github.com/benthomasson/network_ui_test

matburt commented on awx/ui/client/src/network-ui/designs/README.md:
```diff
@@ -0,0 +1,161 @@
+
```
Logs of design docs here that could be condensed into a section in docs/

benthomasson replied:
Moved.

chrismeyersfsu commented on awx/network_ui/urls.py:
```diff
@@ -0,0 +1,10 @@
+
+app_name = 'network_ui'
+urlpatterns = [
+    url(r'^topology.json$', views.json_topology_data, name='json_topology_data'),
```
Tower jobs stdout output is influenced by a GET param `?format=json/html`. I'd rather see 1 url to make future support of the interface easier.

benthomasson replied:
Does Tower support YAML export?

matburt commented on awx/network_ui_test/CONTRIBUTING.md:
```diff
@@ -0,0 +1,221 @@
+
+Network UI Test
```
Is this all just test code? What does it test? This whole module seems to duplicate a lot of code. It's testing but requires its own models, views, migrations, etc? I think I'd prefer a more traditional unit/functional test paradigm that doesn't require a totally different deployment to run.

benthomasson replied:
This code was moved to https://github.com/benthomasson/network_ui_test

matburt commented on awx/network_ui/serializers.py:
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2017 Red Hat, Inc
```
This doesn't use the DRF serializer concepts?

matburt replied:
It's a little hard to understand what is going on and why it's separated from the views if it's not using DRF concepts.

benthomasson replied:
Merged into to views.py

matburt commented on awx/network_ui/serializers.py:
```diff
@@ -0,0 +1,69 @@
+import yaml
+import json
+
+NetworkAnnotatedInterface = Interface.objects.values('name',
```
This is only used one place under `topology_data()` does it need to have module scope?

benthomasson replied:
It's kind of like a database view that's implemented with Django.  It pulls in data from the interface, the connected link, the remote interface, and the remote device.   That's why I used capitalized camel case.

matburt commented on awx/network_ui/views.py:
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2017 Red Hat, Inc
```
This file is strange in that it doesn't use any DRF constructs but maybe that's okay since we've simplified?

I'd merge what you have in serializers down into this... This then becomes a temporary set of views until we make this a much larger feature that requires DRF.

benthomasson replied:
I've merged serializers into views.

chrismeyersfsu reviewed (commented):
My biggest worry is not using DRF serializers.
I don't fully understand why there is duplicate model code for a separate django module testing framework.

benthomasson committed:
Moves network_ui_test to an external repo

chrismeyersfsu commented:
+10 xss resistance

![image](https://user-images.githubusercontent.com/722880/37618701-7a19ed18-2b8d-11e8-8c23-70e829944545.png)

matburt commented on awx/network_ui/models.py:
```diff
@@ -0,0 +1,65 @@
+    name = models.CharField(max_length=200, blank=True)
+    x = models.IntegerField()
+    y = models.IntegerField()
+    id = models.IntegerField()
```
How is this different from `device_id`, could we unify these into the same field?

matburt replied:
In general I'd like to have the pk on the models called `id` to stay consistent with our other models.

benthomasson replied:
I've changed the pk to id on all the models.

matburt commented on awx/network_ui/models.py:
```diff
@@ -0,0 +1,65 @@
+    id = models.IntegerField()
+    device_type = models.CharField(max_length=200, blank=True)
+    interface_id_seq = models.IntegerField(default=0,)
+    host_id = models.IntegerField(default=0,)
```
fk reference to a host?

benthomasson replied:
I've changed host_id and inventory_id to a ForeignKey field.

matburt commented on awx/network_ui/models.py:
```diff
@@ -0,0 +1,65 @@
+    y = models.IntegerField()
+    id = models.IntegerField()
+    device_type = models.CharField(max_length=200, blank=True)
+    interface_id_seq = models.IntegerField(default=0,)
```
What are these `_seq` refrences for?

benthomasson replied:
x_id_seq is used for the next value for a client side assigned id called cid.

jaredevantabor committed:
Disables actions according to RBAC privilege

* Removes the toolbox if user doesn't have permission to edit
* Fixes the extra click that was identified with the context menu
* Adds new readonly version of the move FSM
* Adds an enhancement to debug directive to align the text better
* Disables the toolbox FSM if user doesn't have permission to edit

jaredevantabor committed:
Fixes some bugs from PR feedback

* Fixes bug where new devices on the canvas weren't added to the search dropdown
* Fixes bug with closing the details panel
* Changes the fill color to white for remote-selected devices
* Fixes read-only mode by swapping out move controller for move read-only
* Updates range on the zoom widget

benthomasson committed:
Improves pagination unrolling based on jmcdermott's feedback

* Fixes pagination and fsm-diff
* Removes unused Array.extend

benthomasson committed:
Updates models based on PR feedback from matburt et al.

* Moves topology_data to views
* Changes id to cid
* Changes pk to id
* Changes host_id and inventory_id to ForeignKeys
* Resets migrations for network_ui
* Cleans up old files

benthomasson committed:
Renames Persistence to NetworkingEvents and removes the persistance channel

wwitzel3 commented on Makefile:
```diff
@@ -308,7 +308,7 @@ runworker:
 		. $(VENV_BASE)/awx/bin/activate; \
 	fi; \
-	$(PYTHON) manage.py runworker --only-channels websocket.*
+	$(PYTHON) manage.py runworker --only-channels websocket.* --only-channels persistence
```
Again this name dosen't convey anything useful. Can we call it something like `/networking_events`

benthomasson replied:
I changed Persistence to NetworkingEvents and removed the extra channel completely.

benthomasson replied:
I removed the try/except since that was used for development to generate a browser console log statement when server-side errors happened.  There was no corresponding UI error notification.

wwitzel3 commented on awx/network_ui/routing.py:
```diff
@@ -0,0 +1,10 @@
+    route("websocket.connect", ws_connect, path=r"^/network_ui/topology"),
+    route("websocket.receive", ws_message, path=r"^/network_ui/topology"),
+    route("websocket.disconnect", ws_disconnect, path=r"^/network_ui/topology"),
+    route("persistence", Persistence().handle),
```
`networking_events` instead of `persistence`

benthomasson replied:
I removed this completely in favor of a function call.

wwitzel3 commented on awx/network_ui/consumers.py:
```diff
@@ -0,0 +1,281 @@
+    # Send to all clients editing the topology
+    Group("topology-%s" % message.channel_session['topology_id']).send({"text": message['text']})
+    # Send to persistence handler
+    Channel('persistence').send({"text": message['text'],
```
`networking_events`

benthomasson replied:
I removed this completely in favor of a function call.

wwitzel3 reviewed (changes requested):
Looking good, some suggestions.
